[NP-813] custom table columns add ability to display property of property reference

olhabn opened the pull request:
ticket: https://nanopay.atlassian.net/browse/NP-813
nanopay pr: https://github.com/nanoPayinc/NANOPAY/pull/10121

olhabn committed:
figuring out column view

olhabn committed:
Figuring out view

olhabn committed:
Changing selectedPropertyView

olhabn committed:
Fixing view

olhabn committed:
Fixing warnings

olhabn committed:
array of property names processing added

olhabn committed:
Fixing warnings

olhabn committed:
Fixing views

olhabn committed:
clean up

olhabn committed:
Fixing view

olhabn committed:
Fixing views

olhabn committed:
Fixing views

olhabn committed:
Fixing views

olhabn committed:
Fixing view. deleting comments

olhabn committed:
Fixing views

olhabn committed:
Fixing views

olhabn committed:
Fixing view

olhabn committed:
Fixing view

olhabn committed:
Fixing bug

olhabn committed:
Root property changed from proprety to string

olhabn committed:
Comments deleted

olhabn committed:
using columnsAvailable for all views

olhabn committed:
Fixing view

olhabn committed:
Fixing view

olhabn committed:
Adding arrayview

olhabn committed:
Fixing column arrayView

olhabn committed:
Small fixes

olhabn committed:
Using property label instead of names

olhabn committed:
FIxing error

olhabn committed:
Fixing error

olhabn committed:
Sort labels

olhabn committed:
Some css changes

olhabn committed:
drodown border radius changed

olhabn committed:
Changing selcetedColumns model to string with .

olhabn committed:
Fixing error

olhabn committed:
Figuring out export drivers changes

olhabn committed:
Fixing ColumnConfigToProperty

olhabn committed:
Fixing ColumnConfigToPropertyConverter

olhabn committed:
FIxing userFeedBack to csv

olhabn committed:
Fixing GoogleSheetsOutputter

olhabn committed:
Figuring out fixes for RoleQuery/DashboardAccounts

olhabn committed:
Figuring out drop down

olhabn committed:
Setting correct this for obj. save to localStorage

olhabn committed:
fixing typo

olhabn committed:
do not close on button click

olhabn committed:
Closing on click elsewhere

olhabn committed:
moved dropdown to editcolumnsview

olhabn committed:
Figuring out position

olhabn committed:
figuring out drag and drop

olhabn committed:
Figuring out offset

olhabn committed:
Figuring out sorting

olhabn committed:
figuring out d-d for only seleted columns

olhabn committed:
Moving items on select and un-select

olhabn committed:
fixing isPropertySelected related bugs

olhabn committed:
clean up. fixing bugs

olhabn committed:
fixing bugs

olhabn committed:
fixing padding

olhabn committed:
fixing bugs

olhabn committed:
collapse views on  close

olhabn committed:
figuring out children drag and drop

olhabn committed:
Adding on select drag and drop for

olhabn committed:
Bug fix: isSelectionChanged sub
called on prop change

olhabn committed:
fixing namings and function properties

olhabn committed:
fixing function properties

olhabn committed:
fixing error

olhabn committed:
fixing index assignment

olhabn committed:
fixing d&d

olhabn committed:
fixing error

olhabn committed:
hopefully fixing design mistake

olhabn committed:
fixing columns update on d&d

olhabn committed:
deleting new lines

olhabn committed:
deleting console logs

olhabn committed:
removing unnecessary calls for updating columns

olhabn committed:
adding search and stopPropagation for main div

olhabn committed:
fixing search

olhabn committed:
Fixing logic error

olhabn committed:
fixing error

olhabn committed:
fixing on selection changed

olhabn committed:
mapping properties to its name

olhabn committed:
adding property support for ColumnConfigPropView

olhabn committed:
Fixing things

olhabn committed:
fixing error

olhabn committed:
getting rid of exception Mutations not allowed in OUTPUT state

olhabn committed:
fixing toLowerCase

olhabn committed:
tableColumns added as an option

olhabn committed:
Fixing error

olhabn committed:
adding some comments

olhabn committed:
need to investigate

olhabn committed:
adding come css

olhabn committed:
fixing errors

olhabn committed:
on drag over highlight

olhabn committed:
fixing error

olhabn committed:
clean up

olhabn committed:
clean up

olhabn committed:
Logger added

olhabn committed:
fixing view for Transactions

olhabn committed:
fixing search reset button not working

olhabn committed:
adding padding

olhabn committed:
some bottom margin for search

olhabn committed:
spaces added

olhabn committed:
fixing dashboard

olhabn committed:
deleting comment

olhabn committed:
adding header

olhabn committed:
deleting comment

olhabn committed:
fixing error

olhabn committed:
fixing spacing

olhabn committed:
fixing isInstance

olhabn committed:
fixing error

olhabn committed:
space added

olhabn committed:
fixing formatting

olhabn committed:
merge master

olhabn committed:
fixing "to not export "custom" table columns"
error

olhabn committed:
minor error but still

olhabn committed:
some spacing

olhabn committed:
spaces added

olhabn committed:
spaces added

olhabn committed:
spaces added

olhabn committed:
Merge branch 'master' into NP-813/custom_table_columns_add_ability_to_display_property_of_property

olhabn committed:
figuring out reference

olhabn committed:
Fixing ColumnConfigToPropertyConverter

olhabn committed:
fixing other methods

olhabn committed:
isInstance fixed and
networkTransient check added. not many properties
satisty the check though

olhabn committed:
fixing CSVOutputterImpl java

olhabn committed:
clean up

olhabn committed:
figuring out CSVOutputterImpl fixes

olhabn committed:
fixing google sheets export

olhabn committed:
return added

olhabn committed:
figuring out csvSink

olhabn committed:
figuring out dot and

olhabn committed:
re-writing driver's code

olhabn committed:
using isInstance instead of class id

olhabn committed:
re-arranging imports

olhabn committed:
Moving some methods to TableColumnOutputter

olhabn committed:
adding methods for export

olhabn committed:
fixing errors

olhabn committed:
moving methods

olhabn committed:
fixing filterExportedProps

olhabn committed:
fixing

olhabn committed:
fixing errors

olhabn committed:
removing unnecessary parameter

olhabn commented on src/foam/comics/v2/userfeedback/UserFeedback.js:
```diff
@@ -26,7 +26,15 @@ foam.CLASS({
+        outputter.outputValue("next");
+      `,
+      javaToCSVLabel: `
+      //to remove "infinite" loop
```
this changes were made as PrefixedCSVOutputter was in infinite loop

olhabn committed:
fixing error

olhabn committed:
renaming methods

olhabn committed:
renaming methods

olhabn commented on src/foam/nanos/column_config/ColumnConfigPropertyValue.js:
```diff
@@ -0,0 +1,370 @@
+          return props.filter(p => p.networkTransient).map(p => p.name);
+        } props = [];
+        return propNames.filter(n => { 
+          return allColumnNames.includes(n.split('.')[0]);// && self.returnProperty(of, n).networkTransient;
```
networkTransient is commented out as not a lot of properties are networkTransient.

olhabn replied:
should we set networkTransient default value to true maybe?

jlhughes replied:
I would include 'networkTransient' so that these columns are ignored.

olhabn committed:
fixing error

olhabn committed:
fixing css

olhabn commented on src/foam/u2/view/UnstyledTableView.js:
```diff
@@ -278,10 +307,17 @@ foam.CLASS({
-                var column = typeof axiomOrColumnName === 'string'
-                  ? view.of.getAxiomByName(axiomOrColumnName)
-                  : axiomOrColumnName;
+              forEach(columns_, function([property, overrides]) {
```
Should we use projection instead of retrieving value for a property for UnstyledTableView?

olhabn committed:
deleting unnecessary check

olhabn committed:
clean up

olhabn committed:
clean up

olhabn committed:
clean up

olhabn committed:
fixing errors

olhabn committed:
clean up

olhabn committed:
clean up

olhabn committed:
clean up

olhabn commented on src/foam/nanos/column_config/NestedPropertiesExpression.js:
```diff
@@ -0,0 +1,217 @@
+        Expr e = returnDotExprForNestedProperty(getObjClass(), getNestedProperty().split("\\\\."), 0);
+        if ( e == null )
+          return null;
+        return e.f(obj);
```
this expression context is different from next expression's context  ( which is the same as obj's context ). and this expression context allows to execute find methods but object context doesn't. and I'm not sure what's going on

olhabn committed:
fixing mistake related to StringBuilder

olhabn committed:
merge master

olhabn committed:
fixing errors

olhabn committed:
fixing join

olhabn committed:
fixing css

olhabn committed:
fixing typo

olhabn committed:
networkTransient fixes

olhabn committed:
UserFeedback changes

olhabn committed:
fixing errors

olhabn committed:
fixing things

olhabn committed:
fixing context issue

olhabn committed:
figuring out UnstyledTableView

olhabn committed:
figuring out UnstyledTableView

olhabn committed:
figuring out sort

olhabn committed:
some clean up

olhabn committed:
figuring out bug

olhabn committed:
some exception catching for Dot

olhabn committed:
fixing toString method for projection and
NestedPropertiesExpression as it used in TTLSelectCachingDAO
otherwise change of columns is not reflected

olhabn committed:
fixing slot.get error

olhabn committed:
fixing sort

olhabn committed:
naming fixed

olhabn committed:
fixing naming

olhabn committed:
fixing name

olhabn committed:
fixing duplication

olhabn committed:
fixing columns update

olhabn committed:
fixing dblclick

olhabn committed:
fixing OverlayActionListView

olhabn committed:
fixing currency values

olhabn committed:
fixing image and column update

olhabn committed:
fixing error

olhabn committed:
using tableCellFormatter

olhabn committed:
fixing error

olhabn committed:
clean up

olhabn committed:
fixing formatting

olhabn committed:
maybe justify-content

olhabn committed:
making sort accessible only for 1st level
properties

olhabn committed:
fixing sort

olhabn committed:
changing CSVDriver to CSVTableExportDriver

olhabn committed:
space added

olhabn committed:
fixing returnStringValueForProperty for array

olhabn committed:
fixing things

olhabn committed:
fixing not networkTransient

olhabn committed:
clean up

olhabn committed:
using table column first

olhabn committed:
re-using thisObjValue

olhabn committed:
fixing formatting

olhabn committed:
fixing naming and formatting

olhabn commented on src/foam/nanos/column_config/TableColumnOutputter.js:
```diff
@@ -0,0 +1,110 @@
+          }
+          if ( foam.core.UnitValue.isInstance(prop) ) {
+            if ( unitPropName ) {
+              if ( prop.unitPropName === "destinationCurrency" ) {
```
is there anyway to find targetDAO?

jlhughes replied:
@nanoMichal is there a way to find targetDAO for unitPropName and if there is any place where we map UnitValue to the target DAO, for which unitPropName is primary key?

nanoMichal replied:
No I dont think so. There is no UnitDAO, but we talked about making a unitDAO which would just redirect to the proper dao (currency, security etc) by having a decorator which would check instanceof and direct appropriatly. Would this help?

kgrgreer replied:
Couldn't we use polymorphism to let the property format the value without us having to know the details? Would work for UnitValue but also for the other types as well and would be mor extensible.

olhabn replied:
> Couldn't we use polymorphism to let the property format the value without us having to know the details? Would work for UnitValue but also for the other types as well and would be mor extensible.

I'm not sure if I get it correctly. But my understanding was that unitValue property contains value and fk to unitProp, which stored in different daos. We can use unitValue property the way it is on line 38, but the only object that knows how to format unitValue well is unitProp ( please see line 36 ). 
eg Account has home balance, which is unit value, and it has value, which is 30, and denomination, which is "CAD". Only "CAD" Currency object, retrieved from currency DAO, knows how to format it well.
Let me know if I misunderstood
So I'm not sure about polymorphism and if I get it correctly

olhabn replied:
> No I dont think so. There is no UnitDAO, but we talked about making a unitDAO which would just redirect to the proper dao (currency, security etc) by having a decorator which would check instanceof and direct appropriatly. Would this help?

That would be great. not sure though if we need it that much

olhabn commented on src/foam/u2/view/UnstyledTableView.js:
```diff
@@ -349,157 +409,212 @@ foam.CLASS({
+                  on('mouseover', function() {
+                    if ( !thisObjValue ) {
+                      dao.find(val[0]).then(v => {
+                      thisObjValue = v;
```
I have a problem with context in `find` here (CABankAccount.js) for Bank Accounts and I have some problems with figuring it out. Please let me know if you have any suggestions, I'll keep investigating meanwhile

jlhughes replied:
BankAccounts are still stored in accountsDAO - is that what you were needing?

kgrgreer commented on src/foam/nanos/column_config/TableColumnOutputter.js:
```diff
@@ -0,0 +1,110 @@
+    },
+    {
+      name: 'objectToTable',
+      code: async function(x, of, props, obj) {
```
Why is this async?

olhabn replied:
fixed

kgrgreer replied:
It still says async.

olhabn replied:
it now awaits for another method

olhabn committed:
fixing error

olhabn committed:
clean up

olhabn commented on src/foam/mlang/mlang.js:
```diff
@@ -2696,6 +2701,30 @@ foam.CLASS({
 
         getArray().add(a);
       `
+    },
```
the reason this changes was made is that TTLSelectCachingDAO performs toString on projection for building cache key. and if projection returns only class name on toString method call the cache for the previous projection was used despite the fact that expressions changed

olhabn committed:
reseting CSV related files to master

olhabn committed:
changing CSVTableExportDriver package to export

olhabn committed:
removing CSVDriver

olhabn committed:
Merge branch 'master' into NP-813/custom_table_columns_add_ability_to_display_property_of_property_reference

olhabn committed:
merge master

olhabn committed:
clean up

olhabn committed:
fixing comments

olhabn committed:
header added

olhabn committed:
headers added

olhabn committed:
header fixed

olhabn committed:
new line added

olhabn committed:
fixing error

olhabn committed:
fixing error

olhabn committed:
fixing bug

olhabn committed:
fixing view

olhabn committed:
fixing error

olhabn committed:
fixing error

olhabn committed:
fixing error

olhabn committed:
fixing FObjectProperty or Reference  properties

kgrgreer commented on src/foam/mlang/mlang.js:
```diff
@@ -2696,6 +2701,30 @@ foam.CLASS({
+        var projectionString = this.cls_.name + '(';
+        var exprString = [];
+        for ( var expr of this.exprs ) {
+          exprString.push(expr.toString());
```
I don't think this step is necessary. I think you could just join() this.exprs.

kgrgreer commented on src/foam/mlang/mlang.js:
```diff
@@ -2696,6 +2701,30 @@ foam.CLASS({
+      },
+      javaCode: `
+        StringBuilder sb = new StringBuilder(getClassInfo().getId() + "(");
+        StringJoiner joiner = new StringJoiner(", ");
```
You can also specify the start and end of the string int the StringJoiner which would avoid having to create the StringBuilder and having to copy the data to it.

kgrgreer commented on src/foam/mlang/mlang.js:
```diff
@@ -3240,10 +3277,40 @@ foam.CLASS({
+        FObject obj1;
+        if ( p1 instanceof AbstractFObjectPropertyInfo ) {
+          Object val = getArg1().f(obj);
+          if ( val == null )
```
A bit nicer: return val == null ? : getArg2().f(val);

But your code, so you can choose.

kgrgreer commented on src/foam/mlang/mlang.js:
```diff
@@ -3240,10 +3277,40 @@ foam.CLASS({
+        if ( obj1 == null ) return null;
+        try {
+          return getArg2().f(obj1);
+        } catch( Exception e ) {
```
Would be better to catch Throwable?

kgrgreer commented on src/foam/nanos/column_config/ColumnConfigPropertyValue.js:
```diff
@@ -0,0 +1,223 @@
+              if ( p instanceof foam.core.AbstractFObjectPropertyInfo ) {
+                cls = p.getValueClass();
+              } else {
+                char[] arr = p.getName().toCharArray();
```
On the JS side we have foam.util.String.capitalize(). I think it would make the code easier to understand if we add a similar function to foam.util.StringUtil.java on the java side.

kgrgreer commented on src/foam/nanos/column_config/ColumnConfigPropertyValue.js:
```diff
@@ -0,0 +1,223 @@
+                String s = sb.toString();
+                Method m = ci.getObjClass().getMethod(s, foam.core.X.class);
+                cls = m.getReturnType();
+                sb.setLength(4);
```
What is this line for?

olhabn replied:
it's for StringBuilder clean up and setting it to "find" for next property to use

kgrgreer replied:
Okay, add a comment to make this clear. Thx

kgrgreer commented on src/foam/nanos/column_config/ColumnConfigPropertyValue.js:
```diff
@@ -0,0 +1,223 @@
+    {
+      name: 'returnPropertyAndObject',
+      type: 'foam.nanos.column.ColumnPropertyValue',
+      documentation: `returns property and object on which such property's functions as f or toCSV can be called`,
```
No need for ` when a single-line string without interpolation. Just use ' which parses faster.

kgrgreer commented on src/foam/nanos/column_config/ColumnConfigPropertyValue.js:
```diff
@@ -0,0 +1,223 @@
+                obj1 = (FObject) p.f(obj1);
+                cls = p.getValueClass();
+              } else {
+                char[] arr = p.getName().toCharArray();
```
Create a capitalize() helper function.

kgrgreer commented on src/foam/nanos/column_config/NestedPropertiesExpression.js:
```diff
@@ -0,0 +1,222 @@
+          if ( p instanceof foam.core.AbstractFObjectPropertyInfo ) {
+            cls = p.getValueClass();
+          } else {
+            char[] arr = p.getName().toCharArray();
```
same

olhabn committed:
fixing pr

olhabn committed:
fixing exportDriverRegestrys

olhabn committed:
fixing StringJoiner

olhabn committed:
using throwable

olhabn committed:
adding space

olhabn committed:
isSortable

olhabn committed:
merge CPF-2517/add_isSortable_axiom_to_properties

olhabn committed:
using isSortable

olhabn committed:
fixing property name

olhabn committed:
Merge branch 'CPF-2517/add_isSortable_axiom_to_properties' into NP-813/custom_table_columns_add_ability_to_display_property_of_property_reference

olhabn committed:
changing isSortable to sortable

olhabn committed:
merge master

olhabn committed:
adding comment

olhabn committed:
valueToString added

olhabn committed:
Merge branch 'master' into NP-813/custom_table_columns_add_ability_to_display_property_of_property_reference

jlhughes commented on src/foam/nanos/column_config/CSVTableOutputter.js:
```diff
@@ -0,0 +1,19 @@
+ */
+
+foam.CLASS({
+  package: 'foam.nanos.column',
```
The file system path mush match the package.
please change the directory name to 'column' from column_config

olhabn replied:
fixed

olhabn committed:
replacing column_config with column

kgrgreer commented on src/foam/nanos/column/ColumnConfigPropertyValue.js:
```diff
@@ -0,0 +1,218 @@
+        var allColumnNames = props.map(p => p.name);
+        if ( ! propNames ) {
+          return props.filter(p => ! p.networkTransient ).map(p => p.name);
+        } props = [];
```
This is odd formatting. Why is there a statement on the same line as a }?

olhabn replied:
fixed

kgrgreer commented on src/foam/mlang/mlang.js:
```diff
@@ -2696,6 +2701,22 @@ foam.CLASS({
+    {
+      name: 'toString',
+      code: function() {
+        var projectionString = this.cls_.name + '(' + this.exprs.join(',') + ')';
```
Just combine these two lines since you don't use the var.

olhabn replied:
fixed

kgrgreer commented on src/foam/nanos/export/GoogleSheetsExportDriver.js:
```diff
@@ -30,34 +30,43 @@ foam.CLASS({
+      sheetId = await X.googleSheetsDataExport.createSheet(X, stringArray, metadata);
+      if ( ! sheetId || sheetId.length === 0)
+        return '';
+      var url = `https://docs.google.com/spreadsheets/d/${sheetId}/edit#gid=0`;
```
Could just return.

olhabn replied:
fixed

olhabn committed:
fixing mistakes

olhabn committed:
merge master

olhabn committed:
Merge branch 'master' into NP-813/custom_table_columns_add_ability_to_display_property_of_property_reference

olhabn committed:
hidden: true for CSVTableExportDriver

olhabn committed:
deleting square brackets
